### PR TITLE
RMS Stack Release (Version 0007)

### PR DESCRIPTION
**Pull request by paliwalparitosh**
- Updated Stack UI
  - Simplified Options
  - Adds UI element descriptions
  - New advanced configuration section
- Private OKE cluster support
  - via RMS private endpoint
- Custom helm chart support
  - enables helm upgrade as well
- Logan Entity Creation support
- Auto Log Group Creation
- Adds multiple terraform plan validation checks
- Decouple RMS stack from OKE Monitoring module (main)

### CODE REVIEW

File: terraform/rms_stack/providers.tf [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
```
[the remark this answers is not in the document]
Comment (paliwalparitosh) [reply terraform/rms_stack/providers.tf]: updated in all files.

File: terraform/oke/schema.yaml [hunk 2]
```diff
@@ -20,192 +21,304 @@ locale: "en"
...
-    description: "Tip: To make the log group easy to find in Dashboards and Logs Explorer pages, provide a unique name associated with your cluster name."
+    description: |-
+        To make the log group easy-to-find in Dashboards and Logs Explorer pages, provide a unique name related with your cluster name. 
+        If not provided, the stack creates a log group based on OKE cluster's name and creation date.
     visible:
       and:
-        - opt_create_new_la_logGroup
-    pattern: '^([a-zA-Z0-9]|[a-zA-Z0-9][\\ a-zA-Z0-9_\-]*[\\a-zA-Z\-0-9_])$'  
+        - ${opt_create_new_la_logGroup}
+    pattern: '^([a-zA-Z0-9]|[a-zA-Z0-9][\\ a-zA-Z0-9_\-]*[\\a-zA-Z\-0-9_]$|^$)'
```
[the remark this answers is not in the document]
Comment (paliwalparitosh) [reply terraform/oke/schema.yaml:198]: updated to 

> '(^\S.*$|^$)'

File: terraform/modules/helm/helm.tf [hunk 3]
```diff
@@ -33,10 +38,13 @@ locals {
...
 resource "helm_release" "oci-kubernetes-monitoring" {
   name              = "oci-kubernetes-monitoring"
-  repository        = var.use_local_helm_chart ? null : local.helm_repo_url
-  chart             = var.use_local_helm_chart ? var.helm_abs_path : local.helm_repo_chart
+  repository        = local.repository
+  chart             = local.chart
+  version           = local.version
   wait              = true
   dependency_update = true
+  force_update      = true
```
[the remark this answers is not in the document]
Comment (paliwalparitosh) [reply terraform/modules/helm/helm.tf:46]: We have enabled these 2 properties on helm resource. These seem to make sense to support helm chart upgrade/downgrade via custom helm chart version option.

**force_update:** Force resource update through delete/recreate if needed.
**clean up on fail**: Allow deletion of new resources created in this upgrade when upgrade fails.

File: terraform/oke/schema.yaml [hunk 4]
```diff
@@ -158,74 +145,70 @@ variables:
...
       not:
-        - ${opt_create_new_la_logGroup}
+        - ${opt_create_new_la_log_group}
 
   # New Log Group to collect Kubernetes data
-  oci_la_logGroup_name:
+  oci_la_log_group_name:
     type: string
-    maxLength: 255
+    maxLength: 100 #255
```
Comment (santhoshkvuda) [inline terraform/oke/schema.yaml:173]: remove

File: terraform/oke/stack-required-providers.tf [hunk 5]
```diff
@@ -1,20 +1,20 @@
...
   required_providers {
     oci = {
       source  = "oracle/oci"
-      version = ">= 5.44.0"
+      version = "~> 5.46"
     }
     helm = {
       source  = "hashicorp/helm"
-      version = "2.13.2"
+      version = "~> 2.7.1"
```
Comment (santhoshkvuda) [inline terraform/oke/stack-required-providers.tf:13]: ~> 2.7

File: terraform/oke/version.auto.tfvars [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2023, 2024, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+# TODO: This must be incremented with every release of stack to OCI RMS template
+template_version = "0007"
```
Comment (santhoshkvuda) [inline terraform/oke/version.auto.tfvars:5]: The template version is only to identity the version of template in a particular production region. This version does not control the version of the template to be used by the stack.

File: terraform/oke/stack-inputs.tf [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+# [Released][Hidden] #TODO: Should we retire this input
+# OKE Cluster Name
+variable "oke_cluster_name" {
+  type    = string
+  default = null
+  # User Facing Error
+  validation {
+    condition = var.oke_cluster_name == null ? true : length(regexall("(^\\S+.*$|^$)", var.oke_cluster_name)) > 0
+    #TODO negative and positive test using API
+    error_message = "Invalid input:oke_cluster_name | RegeEx Validation: (^\\S+.*$)"
```
Comment (santhoshkvuda) [inline terraform/oke/stack-inputs.tf:102]: Invalid oke_cluster_name

File: terraform/oke/stack-inputs.tf [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+# OKE Cluster OCID
+variable "oke_subnet_or_pe_ocid" {
+  type    = string
+  default = null
+
+  # User Facing Error
+  validation {
+    condition     = var.oke_subnet_or_pe_ocid == null ? true : length(regexall("^ocid1\\.(subnet|ormprivateendpoint)\\.\\S+$", var.oke_subnet_or_pe_ocid)) > 0
+    error_message = "Incorrect format: var.oke_subnet_or_pe_ocid"
```
Comment (santhoshkvuda) [inline terraform/oke/stack-inputs.tf:134]: Invalid subnet ocid or private endpoint ocid.

File: terraform/oke/stack-inputs.tf [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+# OKE Cluster Entity OCID
+variable "oke_cluster_entity_ocid" {
+  type    = string
+  default = null
+
+  # User Facing Error
+  validation {
+    condition     = var.oke_cluster_entity_ocid == null ? true : length(regexall("^(ocid1\\.loganalyticsentity\\.\\S+)$", var.oke_cluster_entity_ocid)) > 0 ? true : false
+    error_message = "var.oke_cluster_entity_ocid must be set to a valid value."
```
Comment (santhoshkvuda) [inline terraform/oke/stack-inputs.tf:195]: Invalid oke_cluster_entity_ocid

File: terraform/modules/logan/logan.tf [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  log_analytics_entity_id = var.oke_entity_ocid
+  namespace               = local.oci_la_namespace
+
+  lifecycle {
+    # User Facing Error
+    postcondition {
+      # Incorrect Entity OCID check
+      condition     = self.entity_type_name != null
+      error_message = <<-EOT
+        Invalid Entity OCID. Entity does not exist.
```
[the remark this answers is not in the document]
Comment (paliwalparitosh) [reply terraform/modules/logan/logan.tf:33]: updated with

`Authorization failed or requested resource not found.`

File: terraform/oke/schema.yaml [hunk 11]
```diff
@@ -29,10 +29,16 @@ variableGroups:
...
       - ${compartment_ocid}
       - ${current_user_ocid}
-      - ${debug}
       - ${oke_cluster_name} 
     visible: false
 
+  - title: "hidden dev inputs"
+    variables:
+      - ${CLIENT_HOST_OVERRIDES}
+      - ${ENVIRONMENT}
```
Comment (santhoshkvuda) [inline terraform/oke/schema.yaml:38]: remove

File: terraform/oke/schema.yaml [hunk 12]
```diff
@@ -29,10 +29,16 @@ variableGroups:
       - ${boat_tenancy_ocid}
       - ${compartment_ocid}
       - ${current_user_ocid}
-      - ${debug}
       - ${oke_cluster_name} 
     visible: false
```
Comment (santhoshkvuda) [inline terraform/oke/schema.yaml:34]: add a comment

File: terraform/modules/mgmt_agent/format_tags/format_tags.tf [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+#   value = local.freeform_tags
+# }
+
+
+
+# output "defined_tags" {
+#   value = local.defined_tags
+# }
+
+output "defined_tag_list_by_ns" {
```
Comment (santhoshkvuda) [inline terraform/modules/mgmt_agent/format_tags/format_tags.tf:91]: remove

File: terraform/oke/schema.yaml (hunk 11 above)
Comment (paliwalparitosh) [reply terraform/oke/schema.yaml]: removed.

File: terraform/oke/schema.yaml (hunk 12 above)
Comment (paliwalparitosh) [reply terraform/oke/schema.yaml:34]: done.

File: terraform/modules/mgmt_agent/format_tags/format_tags.tf (hunk 13 above)
Comment (paliwalparitosh) [reply terraform/modules/mgmt_agent/format_tags/format_tags.tf]: removed.

**Review by santhoshkvuda — APPROVED** [review]

**Review by pankajtalk — APPROVED** [review]

**Review by palukuri — APPROVED** [review]

**Pull request merged**